presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java

```diff
@@ -49,34 +60,200 @@ public JoinStatsRule(FilterStatsCalculator filterStatsCalculator)
         }
         JoinNode joinNode = (JoinNode) node;
 
-        List<Expression> comparisons = joinNode.getCriteria().stream()
+        PlanNodeStatsEstimate leftStats = lookup.getStats(joinNode.getLeft(), session, types);
+        PlanNodeStatsEstimate rightStats = lookup.getStats(joinNode.getRight(), session, types);
+
+        switch (joinNode.getType()) {
+            case INNER:
+                return Optional.of(computeInnerJoinStats(joinNode, leftStats, rightStats, session, types));
+            case LEFT:
+                return Optional.of(computeLeftJoinStats(joinNode, leftStats, rightStats, session, types));
+            case RIGHT:
+                return Optional.of(computeRightJoinStats(joinNode, leftStats, rightStats, session, types));
+            case FULL:
+                return Optional.of(computeFullJoinStats(joinNode, leftStats, rightStats, session, types));
+            default:
+                return Optional.empty();
+        }
+    }
+
+    private PlanNodeStatsEstimate computeFullJoinStats(JoinNode node, PlanNodeStatsEstimate leftStats, PlanNodeStatsEstimate rightStats, Session session, Map<Symbol, Type> types)
```

Comment:
I would prefer the order of methods to match the order of they appearance in above method